js/util/polygon_poi.js

```diff
@@ -0,0 +1,150 @@
+}
+
+function poleScan(xMin, yMin, xMax, yMax, points, holes, allpoints) {
+    var px, py, pd, maxDistance = 0;
+    // If this is too small then we might miss large areas of the model
+    // If it is too big then we waste cpu searching
+    var splits = 24;
+    // This loop is to prevent some theoretical geometries from failing completely
+    while (maxDistance === 0) {
+        // Sample a grid of points over the shape, starting at the north western edge
+        var yStep = ((yMax - yMin) / splits);
+        var xStep = ((xMax - xMin) / splits);
+        for (var y = yMax; y >= yMin; y -= yStep) {
+            for (var x = xMin; x <= xMax; x += xStep) {
+                // Validate this point is inside the polygon
+                if (isInside(x, y, points)) {
+                    if (holes) {
+                        var inHole = false;
+                        // And it isn't inside a hole
+                        for (var i = 0; i < holes.length && !inHole; i++) {
+                            var hole = holes[i];
+                            inHole = isInside(x, y, hole);
+                        }
+                        if (inHole)
+                            continue;
+                    }
+                    // Work out the distance to the closest perimiter
+                    pd = pointToPerimeterDistance(x, y, allpoints);
+                    if (pd > maxDistance) {
+                        maxDistance = pd;
+                        px = x;
+                        py = y;
+                    }
+                }
+            }
+        }
+        // If we didn't find any points then split the polygon up more.
+        splits = splits * 2;
+    }
+    return {
+        x: px,
+        y: py
+    };
+}
+
+/**
+ * Finds an approximation of a polygon's Pole Of Inaccessibiliy https://en.wikipedia.org/wiki/Pole_of_inaccessibility
+ * Note that for perfect rectangles the pole found ISN'T in the center (though the center is a valid POI)
+ *
+ * @param {Array<Point>} points the outer ring of the polygon
+ * @param {Array<Array<Point>>} [holes] List of interior holes
+ * @param {number} [precision=1] Specified in input coordinate units. If 0 returns after first run, if > 0 repeatedly narrows the search space until the radius of the area searched for the best pole is less than precision
+ *
+ * @returns {Point} Pole of Inaccessibiliy.
+ */
+module.exports = function (points, holes, precision) {
+    if (!isClosedPolygon(points)) {
+        return points[0];
+    }
+
+    if (precision === undefined) {
+        precision = 1;
+    }
+    var xMin, yMin, xMax, yMax;
+    var p = points[0];
+    xMin = xMax = p.x;
+    yMin = yMax = p.y;
+    for (var i = 1; i < points.length; i++) {
+        p = points[i];
+        if (xMin > p.x) xMin = p.x;
+        if (xMax < p.x) xMax = p.x;
+        if (yMin > p.y) yMin = p.y;
+        if (yMax < p.y) yMax = p.y;
+    }
+
+    var allpoints = holes ? points.concat.apply(points, holes) : points;
```

Comment:
By reorganizing the code a bit, I think we can fully avoid doing `concat` (which copies all the data), gaining a bit of performance.